src/Search/Search.Tests/ModelComparer/ModelComparer.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+namespace Microsoft.Azure.Search.Tests.Utilities
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Compares instances of a type for structural equality, taking into account public properties, collections, and dictionaries.
+    /// </summary>
+    /// <typeparam name="T">Type of objects to compare.</typeparam>
+    /// <remarks>
+    /// <para>
+    /// Supported types include enums, primitives, nullables, List, Dictionary, and DTOs having public properties
+    /// composed of the supported types (including other nested DTOs). The rules are different for different types.
+    /// </para>
+    /// <para>
+    /// Any type implementing IEnumerable, including dictionaries, are compared element by element and compose with the other mapping
+    /// rules on element type.
+    /// </para>
+    /// <para>
+    /// DTOs are compared by the values of their public properties. Note that comparisons are performed according to the actual most-derived
+    /// type of the objects (which must match), not the apparent static type of the references.
+    /// </para>
+    /// <para>
+    /// Enums, primitive types, and any other type not falling into the other categories are compared using Object.Equals().
+    /// </para>
+    /// </remarks>
+    public class ModelComparer<T> : IEqualityComparer<T>
+    {
+        private IEqualityComparer _comparer = new DynamicModelComparer(typeof(T));
+
+        public bool Equals(T x, T y) => _comparer.Equals(x, y);
+
+        public int GetHashCode(T obj) => obj.GetHashCode();
+
+        private class DynamicModelComparer : IEqualityComparer
+        {
+            private Type _type;
+
+            public DynamicModelComparer(Type type)
+            {
+                _type = type;
+            }
+
+            int IEqualityComparer.GetHashCode(object obj) => obj.GetHashCode();
+
+            bool IEqualityComparer.Equals(object x, object y)
+            {
+                if (_type.CanBeNull() && x == null || y == null)
```

Comment:
Seems pretty obvious to me. `&&` and `||` always have lower precedence than `==`